PDEP-1: Purpose and guidelines for pandas enhancement proposals

datapythonista opened the pull request:
Closes #28568

Initial PDEP to define purpose and guidelines for pandas enhancement proposals (equivalent to PEPs or NEPs). **Feedback very welcome**.

This PR also makes the PDEPs public in the [roadmap page](https://pandas.pydata.org/about/roadmap.html) of our website.

datapythonista committed:
PDEP-1: Purpose and guidelines for pandas enhancement proposals

datapythonista committed:
black

datapythonista committed:
Update PR number

simonjayhawkins commented on web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md:
```diff
@@ -0,0 +1,109 @@
+
+Anyone can propose a PDEP, but in most cases developers of pandas itself and related
+projects are expected to author PDEPs. If you are unsure if you should be opening
+an issue or creating a PDEP, it's probably save to start by
```
```suggestion
an issue or creating a PDEP, it's probably safe to start by
```

simonjayhawkins commented on web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md:
```diff
@@ -0,0 +1,109 @@
+# PDEP-1: Purpose and guidelines
+
+- Date 21 June 2022
```
```suggestion
- Date: 21 June 2022
```

datapythonista committed:
Update web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md

Co-authored-by: Simon Hawkins <simonjayhawkins@gmail.com>

datapythonista committed:
Update web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md

Co-authored-by: Simon Hawkins <simonjayhawkins@gmail.com>

mroeschke commented:
This would probably close https://github.com/pandas-dev/pandas/issues/28568

mroeschke commented on web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md:
```diff
@@ -0,0 +1,109 @@
+A PDEP will be accepted by the core development team, and decisions will be made
+based on the [pandas governance document](https://github.com/pandas-dev/pandas-governance/blob/master/governance.md).
+
+Once a PDEP is accepted, it is unclear how the implementation will happen. The
```
```suggestion
Once a PDEP is accepted, any contributions can be made toward the implementing the PDEP with an open-ended completion timeline . The
```

IMO the former wording doesn't sound that encouraging to propose PDEPs that might not be implemented

mroeschke commented on web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md:
```diff
@@ -0,0 +1,109 @@
+
+#### Implemented PDEP
+
+Once a PDEP is implemented and available in the main branch of pandas, its
```
Should the release version be included too to note which version of pandas has the complete implementation?

datapythonista replied:
Great idea, I included it.

datapythonista committed:
Update web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md

Co-authored-by: Matthew Roeschke <emailformattr@gmail.com>

datapythonista committed:
Implemented PDEPs including pandas version

datapythonista committed:
Merge branch 'pdep' of github.com:datapythonista/pandas into pdep

mroeschke reviewed (commented):
I didn't immediately see this, but it would be good to explain how an accepted/rejected decision is made. This looks to be Numpy's: https://numpy.org/neps/nep-0000.html#how-a-nep-becomes-accepted

For example, it would be good to define:

* How consensus is defined: 100%, 2/3rds, 51% approval from core devs?
* How to collect consensus: PDEP PR? Monthly dev meeting?
* Should a "call to vote" action be available to unblock perpetual discussion?

datapythonista commented:
> I didn't immediately see this, but it would be good to explain how an accepted/rejected decision is made.

Very good point. I thought this was quite a big topic, and maybe more related to general governance than specific to PDEPs. That's why for now I decided to just link to the governance documentation. Based on the current governance, decisions are made by consensus, with the BDFL (Wes) having the final word when consensus is not possible.

I think the governance documentation is 7 years old. Many things changed since then, including the structure of the team (from like 3 active core devs to like 20, Python moved out of a BDFL model...). Also, besides the decision making, I think our governance seems outdated. I personally don't know most of the people in the pandas CoC committee, and I assume they are not performing any task, and probably don't even know they are part of it.

While I fully agree on what you say, I'd personally keep the discussions separate, as I think both are quiet big, and can be taken care of separately. What do you think?

mroeschke reviewed (approved):
Sounds good to me. Agreed that general decision making details can go under governance.

lithomas1 commented on web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md:
```diff
@@ -0,0 +1,110 @@
+## Evolution of this PDEP
+
+While most PDEPs aren't expected to change after accepted, in some cases like this
+PDEP, they will be updated to contain its latest version, if things evolve. A log
```
We should probably clarify when a PDEP can be revised v.s. when a new PDEP has to be submitted superseding the old PDEP.

lithomas1 commented on web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md:
```diff
@@ -0,0 +1,110 @@
+The file is a markdown file, you can use `web/pdeps/accepted/0001.md` as a reference
+for the expected format.
+
+By default, we expect a PDEP will be accepted, so the PR of a PDEP should be done
```
Do we want to also have a public comment period on PDEP's like tensorflow has with their RFC's?

We should probably also clarify when voting happens(define what proportion of core team members need to consider a PDEP ready before voting like @mroeschke stated) and how long core team members have to vote.

simonjayhawkins replied:
> But again, I'd start by seeing how things work, and if we see PDEPs keep open for too long, and seems like having a timeframe should help, surely worth giving it a try.

I agree, just having a PDEP is akin to a new process, so having a process for the PDEP is maybe like introducing another meta process at the same time. So in response to this comment and others related to the process, I agree with @datapythonista that to begin with we just discuss whether we want to use PDEPs, what we want out of a PDEP,  and roughly what it should contain and iterate on the gaps over time. The only caveat here, is that once someone has spent effort preparing a PDEP and it is approved that we don't have "blockers" implementing, so that does affect the approval process to some degree.

jreback replied:
i agree - we need a voting process here - we could emulate that of NEP which is pretty simple

jorisvandenbossche replied:
> we need a voting process here

I think what Marc proposed above is to defer a discussion about the decision process (basically our governance model) for a subsequent discussion.

(which sounds good to me to keep the discussion manageable)

lithomas1 reviewed (commented):
LG overall. Just had a few questions as I was going through it.

datapythonista commented on web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md:
```diff
@@ -0,0 +1,110 @@
+#### Submitting a PDEP
+
+Proposing a PDEP is done by creating a PR adding a new file to `web/pdeps/accepted/`.
+The file is a markdown file, you can use `web/pdeps/accepted/0001.md` as a reference
```
I'd personally leave that for later, once we need it. Those sound like good ideas to me, but I wouldn't implement them initially, I would wait to see how things work first. If we merge like PDEP per month, I think checking the last PDEP number before merging is easier than a system to autogenerate them. And for a template, I'd wait to have few actual PDEPs before deciding if it helps, or if PDEPs are too different from one to another.

Does it make sense to you to start simple and iterate later as we have more experience?

datapythonista commented on web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md:
```diff
@@ -0,0 +1,110 @@
+for the expected format.
+
+By default, we expect a PDEP will be accepted, so the PR of a PDEP should be done
+in the `accepted` directory and contain `Status: Accepted`. If a PDEP is finally
```
Sounds good to me.

Regarding merging PDEPs as draft, feels like it could make discussions more difficult. Like where to have the discussion once the PR is merged. Via new PRs? Or in a second PR, not being able to comment in parts not in the diff. But I guess in some cases it can be useful, maybe a PR with different parts, and merging as a Draft once the first part has been discussed, and before working on other parts.

I'll update the document if there are no objections to using Under discussion status (or Draft if people prefer this name) until PR is Accepted. And to consider merging PRs before they are accepted when it's useful.

jorisvandenbossche replied:
> Regarding merging PDEPs as draft, feels like it could make discussions more difficult.

We will probably have to see how it goes in practice. For many PDEPs a single PR with proposal, discussion, acceptance all in one will quite likely be sufficient. 
But I can imagine for larger/longer-running discussions (I am thinking eg about the copy/view proposal), that it could be useful to already have a hosted version of PDEP that is still under discussion. 

As some references: numpy seems to already merge NEP PRs if they are still under discussion (eg https://github.com/numpy/numpy/pull/18456), but they also have substantial discussion on the mailing list. Similarly in Python, PEPs are merged quickly, but also there the discussion is mostly happening on mailing list or discourse. That's of course different with the current way in pandas where most discussion typically happens on GitHub.

rhshadrach commented on web/pandas/about/roadmap.md:
```diff
@@ -15,10 +15,53 @@ fundamental changes to the project that are likely to take months or
+[PEDP-1](/pdeps/accepted/0001-puropose-and-guidelines.html).
 
-## Extensibility
+## PDEPs
```
I know this is bikeshedding, and certainly okay with it as-is, but maybe the abbreviation should be pdEP? My thinking here is:

 - With the "p" in pandas traditionally lowercase; as opposed to the P in Python (PEP) and N in NumPy (NEP)
 - Reflects `import pandas as pd`
 - I instinctively see "PDEP" and "P-DEP"

datapythonista replied:
I'd personally try to avoid adding extra complexity with the capitalization. I don't know what P-DEP is or what stands for. If it's something we want to avoid, my preference would be to use something totally different. Like PCP or PDCP (C for Change instead of E for ENHANCEMENT), or anything else. But if people prefer pdEP it, I'll update it.

jreback replied:
why not just PEP ; i don't think this conflicts with python enhancements and it's simpler and easier to pronounce

bashtage replied:
PEP could be hard to find using search engines given the ubiquity of PEP as Python Enhancement Proposal.

datapythonista committed:
Merge remote-tracking branch 'upstream/main' into pdep

datapythonista committed:
Merge remote-tracking branch 'upstream/main' into pdep

datapythonista committed:
Addressed feedback from reviews, couple of visualization fixes, and styles improvements

datapythonista commented:
I updated the PR to incorporate the feedback from comments, and I replied to the ones I didn't directly incorporate.

I also fixes couple of visualization issues, and improved a bit the styles, so the roadmap and also other pages are clearer.

Let me know of additional feedback, and feel free to approve when you're happy with getting this merged, or add a `Request changes` flagged if you see any blocker. Thanks!

datapythonista commented:
@pandas-dev/pandas-core any other comment before merging this? If there are no objections I'll merge this in couple of days, and we can keep iterating as needed in separate PRs.

jreback commented on web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md:
```diff
@@ -0,0 +1,121 @@
+The file is a markdown file, you can use `web/pdeps/accepted/0001.md` as a reference
+for the expected format.
+
+By default, we expect a PDEP will be accepted, so the PR of a PDEP should be done
```
i am not sure this is true - sure a lot will easily be accepted but some might be controversial and ultimately not accepted

simonjayhawkins replied:
yeah. maybe just qualify this by changing `By default, we expect a PDEP will be accepted` to something like `By default, we expect a PDEP (with some preliminary discussion) will be accepted` 

before this text we have...

> A PDEP should be used for changes that are not
immediate and not obvious, and are expected to require a significant amount of
discussion and require detailed documentation before being implemented.

so the PEP is part of the detailed documentation and an issue should perhaps be the initial part of the discussion. (just as I would normally expect say a bug fix PR to have an associated issue)

and we also have 

> If you are unsure if you should be opening
an issue or creating a PDEP, it's probably safe to start by
[opening an issue](https://github.com/pandas-dev/pandas/issues/new/choose), which can
be eventually moved to a PDEP.

I think that maybe issues should **always** be opened before submitting a PEP, either as a specific issue or an existing issue concluding that a PEP is required. Maybe we also need to ensure that `@pandas-dev/pandas-core` is also notified in these cases.

If we have some form of gate for opening a PEP, then most should be accepted.

simonjayhawkins reviewed (approved):
Thanks @datapythonista generally lgtm.

jorisvandenbossche reviewed (commented):
It might be useful to somewhere list the possible "status" values? (now those seem to be included in the different workflow sections, but it's not very explicit)

rhshadrach commented on web/pandas/pdeps/accepted/0001-purpose-and-guidelines.md:
```diff
@@ -0,0 +1,121 @@
+#### Rejected PDEP
+
+A PDEP can be rejected when the final decision is that its implementation is
+not the best for the interests of the project. They are as useful as accepted
```
This is to contrast rejected PDEPs against accepted PDEPs; i.e. "even rejected PDEPs are useful". I think the wording is okay, but maybe "Rejected PDEPs are just as useful as accepted PDEPs..." would make it more clear?

jorisvandenbossche replied:
Ah, I missed the first "as" in the sentence, so I read it as "they are useful as accepted PDEP", instead of "they are _as_ useful as accepted PDEPs", hence the confusion.

jorisvandenbossche commented:
Related to my previous comment (https://github.com/pandas-dev/pandas/pull/47444#discussion_r905868612) about starting a new PDEP as draft / under discussion: this has been changed in the text, but at the same time we still say to create the new PDEP file in the ``web/pdeps/accepted/`` directory. 
Personally, I still find that a bit confusing and sending a potentially wrong message. If we have an `under_discussion` status, why not also reflect that in the directory / path? (although this probably also comes back to whether we have the intention to ever merge PRs in that status, because indeed if we never do that, we would actually never use that directory in practice)

Taking one step back: do we actually need those different directories? (I suppose they are currently used to simplify the script to automate the website content to list the PDEPs) 
For example, NumPy (and also Python PEPs) uses a flat directory for the documents: https://github.com/numpy/numpy/tree/main/doc/neps. And it then infers the status from the header. If we have a standard preamble that has the status in that list, that also sounds doable to automate this (instead of being based on directories)? It would make my above discussion point (about in which directory to start the proposal) moot, and can also simplify the overall process a little bit (eg you never need to move files for changing a status, but just change the Status line in the document (eg when a PDEP gets "implemented")).

datapythonista commented:
Thanks all for the feedback, very interesting points.

For the voting, I'm working on the governance documents in parallel, and I'll propose changes there to have a way to decide when a PDEP is ready to merge.

For the name to use, there are several comments about names, and many valid points. It's probably not so important, but to find something that most people are happy with, I think we can run a quick poll. I'll write here different options including the ones already proposed. Feel free to suggest more, and when we're ready I'll send a link to the pandas-dev list with a poll.

- PDEP
- pdEP
- PEP
- EP
- PCP (pandas change proposal)
- DEP (dataframe enhancement proposal)

I'll be addressing the rest of the comments in this PR.

noatamir commented:
May I suggest adding a template for new proposals as part of this PR, or as a follow up if that's preferable? 😇 

I find templates useful to help facilitate structured discussions, and that important points relevant for consideration aren't missed by the proposal author(s) before submitting them, e.g. how would this affect backwards compatibility.  

For reference I include these two templates which could be relevant as a starting point:
- scikit-learn's [SLEP template](https://scikit-learn-enhancement-proposals.readthedocs.io/en/latest/slep_template.html)
- NumPy's [NEP template](https://numpy.org/neps/nep-template.html)

datapythonista committed:
Merge main

datapythonista committed:
Addressing comments from reviews

datapythonista commented:
Thanks all for the feedback. Seems like the preferred name is PDEP, so keeping it. I addressed all the comments, there are two pending things, that I think make more sense to address in future PRs:

- How the exact decision on when a PDEP is approved (which is part of a wider discussion about governance)
- Adding a template for PDEPs. While this seems a great idea, I personally think it's premature to have it initially, and I'd wait until we've got few PDEPs, and then decide if we think it adds value, and what should be the format, with concrete examples

Let me know if any other comment. @jreback you reviewed with `Requested changes`, I rewrote that whole section that other people also found misleading, I think your comments should be addressed now.

simonjayhawkins commented on web/pandas/pdeps/0001-purpose-and-guidelines.md:
```diff
@@ -0,0 +1,127 @@
+
+Once a PDEP is implemented and available in the main branch of pandas, its
+status will be changed to `Status: Implemented`, so there is visibility that the PDEP
+is not part of the roadmap and future plans, but a change that it already
```
```suggestion
is not part of the roadmap and future plans, but a change that has already
```

simonjayhawkins commented on web/pandas/pdeps/0001-purpose-and-guidelines.md:
```diff
@@ -0,0 +1,127 @@
+status will be changed to `Status: Implemented`, so there is visibility that the PDEP
+is not part of the roadmap and future plans, but a change that it already
+happened. The first pandas version in which the PDEP implementation is
+available will also be included in the PDEP.
```
not yet relevant, but would this be just a note in the revision history or something else?

simonjayhawkins commented on web/pandas/pdeps/0001-purpose-and-guidelines.md:
```diff
@@ -0,0 +1,127 @@
+#### Rejected PDEP
+
+A PDEP can be rejected when the final decision is that its implementation is
+not the best for the interests of the project. Rejected PDEPs are as useful as accepted
```
```suggestion
not in the best interests of the project. Rejected PDEPs are as useful as accepted
```

simonjayhawkins commented on web/pandas/pdeps/0001-purpose-and-guidelines.md:
```diff
@@ -0,0 +1,127 @@
+
+For submitted PDEPs that do not contain proper documentation, are out of scope, or
+are not useful to the community for any other reason, the PR will be closed after
+discussion with the author, instead of merging them as rejected. This is to not
+add noise to the list of rejected PDEPs, which should contain documentation as
```
```suggestion
discussion with the author, instead of merging them as rejected. This is to avoid
adding noise to the list of rejected PDEPs, which should contain documentation as
```

simonjayhawkins reviewed (approved):
Thanks @datapythonista

jreback commented on web/pandas/about/roadmap.md:
```diff
@@ -15,10 +15,35 @@ fundamental changes to the project that are likely to take months or
+
+<div class="alert alert-warning" role="alert">
+  pandas is in the process of moving roadmap points to PDEPs (implemented in
+  June 2022). During the transition, some roadmap points will exist as PDEPs,
```
yeah switch to August :->

jreback reviewed (approved):
lgtm. some open comments by @simonjayhawkins

rhshadrach reviewed (approved):
Regarding putting up PDEPs under discussion as open PRs, the one negative aspect is that they aren't protected. The author can force-push making changes hard to follow. I think this will work well assuming no bad actors, and I think that's an okay assumption to make, but wanted to be explicit about it. Also, this downside is outweighed by the upside of being able to have all discussion in a consolidated place in my opinion.

datapythonista committed:
Update web/pandas/pdeps/0001-purpose-and-guidelines.md

Co-authored-by: Simon Hawkins <simonjayhawkins@gmail.com>

datapythonista committed:
Update web/pandas/pdeps/0001-purpose-and-guidelines.md

Co-authored-by: Simon Hawkins <simonjayhawkins@gmail.com>

datapythonista committed:
Update web/pandas/pdeps/0001-purpose-and-guidelines.md

Co-authored-by: Simon Hawkins <simonjayhawkins@gmail.com>

datapythonista committed:
Update web/pandas/pdeps/0001-purpose-and-guidelines.md

Co-authored-by: Simon Hawkins <simonjayhawkins@gmail.com>

datapythonista committed:
Last review comments and dates updated

datapythonista commented:
Thanks all for the reviews and feedback!

h-vetinari commented:
An unexpected & pleasant surprise that this old issue of mine got fixed. Not sure I have the time/energy to pour those old PDEP-sized ideas into documents, but glad to know that there'd now be a way to do it.

simonjayhawkins commented:
Thanks @datapythonista for the PR.

> An unexpected & pleasant surprise that this old issue of mine got fixed. Not sure I have the time/energy to pour those old PDEP-sized ideas into documents, but glad to know that there'd now be a way to do it.

Thanks @h-vetinari. feel free to comment on those issues that a PDEP may be a good idea as others may think that they are worth pushing now.

simonjayhawkins commented:
and if you have already put a lot of time and effort in the discussion, you may have already made it alot easier to progress.
